talk/morelanguage/random.tex

```diff
@@ -0,0 +1,81 @@
+        \item distributions to shape these into numbers
+        \item access to (potential) hardware entropy
+      \end{itemize}
+      \item Found in header \mintinline{cpp}{<random>}
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+
+ \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Engines}
+  \begin{block}{Engines}
+    \begin{itemize}
+      \item Generate random bits (as integers), depending on algorithm
+      \item Can be seeded via constructor or \mintinline{cpp}{seed()}
+      \item Algorithms: \mintinline{cpp}{linear_congruential_engine}, \mintinline{cpp}{mersenne_twister_engine}, \mintinline{cpp}{subtract_with_carry_engine}
+      \item Adaptors: \mintinline{cpp}{discard_block_engine}, \mintinline{cpp}{independent_bits_engine}, \mintinline{cpp}{shuffle_order_engine}
+      \item Aliases (use these): \mintinline{cpp}{minstd_rand0}, \mintinline{cpp}{minstd_rand}, \mintinline{cpp}{mt19937}, \mintinline{cpp}{mt19937_64}, \mintinline{cpp}{ranlux24}, \mintinline{cpp}{ranlux48}, \mintinline{cpp}{knuth_b}, \mintinline{cpp}{default_random_engine}
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    int seed = ...;
+    std::default_random_engine engine{seed};
+    auto i = engine(); // operator(): random integer
+ \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Random device}
+  \begin{block}{Random device}
+    \begin{itemize}
+      \item PRNGs will produce the same numbers for the same seed
+      \item PRNGs are \emph{deterministic} (may be useful or not)
+      \item To get different numbers on each run, we must provide a truly random number as seed
+      \item \mintinline{cpp}{std::random_device} provides \emph{non-deterministic} random numbers
+      \item It uses hardware provided entropy, if available
+      \item It may become slow when called too often (e.g. when hardware entropy pool is exhausted)
+      \item Use it only to seed a PRNG!
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    std::random_device rd;
+    auto seed = rd();
+    std::default_random_engine engine{seed};
+ \end{cppcode}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Distributions}
+  \begin{block}{Distributions}
+    \begin{itemize}
+      \item Take random bits from engines and shape them into numbers
+      \item Standard library provides a lot of them, including uniform, bernoulli, poisson, normal and sampling distributions
+      \item See \href{https://en.cppreference.com/w/cpp/numeric/random}{cppreference} for the full list
+    \end{itemize}
+  \end{block}
+  \begin{cppcode}
+    std::random_device rd;
+    std::default_random_engine engine{rd()};
+
+    std::normal_distribution<double> dist(5.0, 1.5); // µ, σ
+    double r = dist(engine); // random normally dist. value
+
+    std::uniform_int_distribution<int> dist2(0, 10);
+    std::vector<int> v(100);
+    for (auto& e : v)
+      e = dist2(engine);
+ \end{cppcode}
```

Comment:
```suggestion
  \begin{exampleblock}{Example code}
  \begin{cppcode}
    std::random_device rd;
    std::default_random_engine engine{rd()};

    std::normal_distribution<double> dist{5.0, 1.5}; // µ, σ
    double r = dist(engine); // random normally dist. value

    std::uniform_int_distribution<int> dist2{0, 10};
    std::vector<int> v(100);
    std::generate(begin(v), end(v), [&](){ return dist2(engine); });
 \end{cppcode}
 \end{exampleblock}
```

Reply:
And drop empty lines if it does not fit, or go to smaller font

Reply:
`std::generate` is a good idea!